File: airi/apps/stage-pocket/android/app/src/main/java/ai/moeru/airi_pocket/websocket/HostWebSocketBridge.kt
package ai.moeru.airi_pocket.websocket

import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

interface HostWebSocketSession {
    fun send(text: String)
    fun close(code: Int?, reason: String?)
}

sealed interface HostWebSocketEvent {
    data object Open : HostWebSocketEvent
    data class Message(val text: String) : HostWebSocketEvent
    data class Error(val message: String) : HostWebSocketEvent
    data class Close(val code: Int?, val reason: String?) : HostWebSocketEvent
}

class HostWebSocketBridge(
    private val sessionFactory: (url: String, emit: (HostWebSocketEvent) -> Unit) -> HostWebSocketSession,
    private val eventSink: (payload: String) -> Unit,
) {
    private val sessions = ConcurrentHashMap<String, HostWebSocketSession>()

    fun handleCommand(payload: String) {
        val command = JSONObject(payload)
        when (command.getString("kind")) {
            "connect" -> handleConnect(command)
            "send" -> handleSend(command)
            "close" -> handleClose(command)
        }
    }

    fun dispose() {
        sessions.values.forEach { it.close(null, "Bridge disposed") }
        sessions.clear()
    }

    private fun handleConnect(command: JSONObject) {
        val id = command.getString("id")
        runCatching {
            sessionFactory(command.getString("url")) { event -> handleSessionEvent(id, event) }
        }
            .onSuccess { sessions[id] = it }
            .onFailure { error ->
                val message = error.message ?: "Failed to create websocket session"
                emitEvent(kind = "error", id = id) {
                    put("message", message)
                }
                emitEvent(kind = "close", id = id) {
                    put("reason", message)
                }
            }
    }

    private fun handleSend(command: JSONObject) {
        val id = command.getString("id")
        sessions[id]?.send(command.getString("data"))
    }

    private fun handleClose(command: JSONObject) {
        val id = command.getString("id")
        sessions[id]?.close(command.optNullableInt("code"), command.optNullableString("reason"))
    }

    private fun handleSessionEvent(id: String, event: HostWebSocketEvent) {
        when (event) {
            HostWebSocketEvent.Open -> emitEvent(kind = "open", id = id)
            is HostWebSocketEvent.Message -> emitEvent(kind = "message", id = id) {
                put("data", event.text)
            }
            is HostWebSocketEvent.Error -> emitEvent(kind = "error", id = id) {
                put("message", event.message)
            }
            is HostWebSocketEvent.Close -> {
                sessions.remove(id)
                emitEvent(kind = "close", id = id) {
                    event.code?.let { put("code", it) }
                    event.reason?.let { put("reason", it) }
                }
            }
        }
    }

    private fun emitEvent(
        kind: String,
        id: String,
        build: JSONObject.() -> Unit = {},
    ) {
        eventSink(
            JSONObject()
                .put("kind", kind)
                .put("id", id)
                .apply(build)
                .toString(),
        )
    }
}

private fun JSONObject.optNullableInt(key: String): Int? =
    if (has(key) && !isNull(key)) getInt(key) else null

private fun JSONObject.optNullableString(key: String): String? =
    if (has(key) && !isNull(key)) getString(key) else null

File: airi/apps/stage-pocket/ios/App/App/HostWebSocketBridge.swift
import Foundation

protocol HostWebSocketSession {
    func send(text: String)
    func close(code: Int?, reason: String?)
}

enum HostWebSocketEvent {
    case open
    case message(String)
    case error(String)
    case close(Int?, String?)
}

private struct HostBridgeCommand: Decodable {
    let kind: String
    let id: String
    let url: String?
    let data: String?
    let code: Int?
    let reason: String?
}

private struct HostBridgeEventPayload: Encodable {
    let kind: String
    let id: String
    let data: String?
    let message: String?
    let code: Int?
    let reason: String?
}

final class HostWebSocketBridge {
    typealias SessionFactory =
        (_ url: String, _ emit: @escaping (HostWebSocketEvent) -> Void) throws -> HostWebSocketSession

    private let sessionFactory: SessionFactory
    private let eventSink: (String) -> Void
    private var sessions: [String: HostWebSocketSession] = [:]
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(
        sessionFactory: @escaping SessionFactory,
        eventSink: @escaping (String) -> Void
    ) {
        self.sessionFactory = sessionFactory
        self.eventSink = eventSink
    }

    func handleCommand(_ payload: String) {
        guard let data = payload.data(using: .utf8) else {
            return
        }

        guard let command = try? decoder.decode(HostBridgeCommand.self, from: data) else {
            return
        }

        DispatchQueue.main.async {
            switch command.kind {
            case "connect":
                self.handleConnect(command)
            case "send":
                self.handleSend(command)
            case "close":
                self.handleClose(command)
            default:
                break
            }
        }
    }

    func dispose() {
        DispatchQueue.main.async {
            self.sessions.values.forEach { $0.close(code: nil, reason: "Bridge disposed") }
            self.sessions.removeAll()
        }
    }

    private func handleConnect(_ command: HostBridgeCommand) {
        guard let url = command.url else {
            emitEvent(kind: "error", id: command.id, message: "Missing websocket url")
            emitEvent(kind: "close", id: command.id, reason: "Missing websocket url")
            return
        }

        print("[HostWebSocketBridge] connect id=\(command.id) url=\(url)")

        do {
            let session = try sessionFactory(url) { [weak self] event in
                self?.handleSessionEvent(id: command.id, event: event)
            }
            sessions[command.id] = session
        } catch {
            print("[HostWebSocketBridge] connect failed id=\(command.id) error=\(error.localizedDescription)")
            let message = error.localizedDescription
            emitEvent(kind: "error", id: command.id, message: message)
            emitEvent(kind: "close", id: command.id, reason: message)
        }
    }

    private func handleSend(_ command: HostBridgeCommand) {
        guard let data = command.data else {
            return
        }

        sessions[command.id]?.send(text: data)
    }

    private func handleClose(_ command: HostBridgeCommand) {
        sessions[command.id]?.close(code: command.code, reason: command.reason)
    }

    private func handleSessionEvent(id: String, event: HostWebSocketEvent) {
        DispatchQueue.main.async {
            print("[HostWebSocketBridge] event id=\(id) \(String(describing: event))")
            switch event {
            case .open:
                self.emitEvent(kind: "open", id: id)
            case .message(let data):
                self.emitEvent(kind: "message", id: id, data: data)
            case .error(let message):
                self.emitEvent(kind: "error", id: id, message: message)
            case .close(let code, let reason):
                self.sessions.removeValue(forKey: id)
                self.emitEvent(kind: "close", id: id, code: code, reason: reason)
            }
        }
    }

    private func emitEvent(
        kind: String,
        id: String,
        data: String? = nil,
        message: String? = nil,
        code: Int? = nil,
        reason: String? = nil
    ) {
        let payload = HostBridgeEventPayload(
            kind: kind,
            id: id,
            data: data,
            message: message,
            code: code,
            reason: reason
        )

        guard let encoded = try? encoder.encode(payload),
              let string = String(data: encoded, encoding: .utf8) else {
            return
        }

        eventSink(string)
    }
}
